bundles/org.eclipse.cdt.lsp/src/org/eclipse/cdt/lsp/server/enable/HasLanguageServerPropertyTester.java

```diff
@@ -26,9 +38,49 @@ public HasLanguageServerPropertyTester() {
 	@Override
 	public boolean test(Object receiver, String property, Object[] args, Object expectedValue) {
 		if (cLanguageServerProvider != null) {
-			return cLanguageServerProvider.isEnabledFor(receiver);
+			if (receiver instanceof URI) {
+				// called from the language server enabler for LSP4E:
+				var uri = (URI) receiver;
+				var fileHandle = getFileHandle(uri);
+				if(fileHandle != null && fileHandle instanceof IResource) {
+					var resource = (IResource) fileHandle;
+					if (resource != null && resource.getProject() != null) {					
+						return cLanguageServerProvider.isEnabledFor(resource.getProject());
+					}
+				}
+				// when resource == null it's an external file: Check if the file is already opened, if not check the active editor:
+				return LspUtils.isFileOpenedInLspEditor(uri);
+			} else if (receiver instanceof ITranslationUnit) {
+				// called to enable the LS based CSymbolsContentProvider: 
+				return Optional.of((ITranslationUnit) receiver)
+						 .map(ITranslationUnit::getCProject)
+						 .map(ICProject::getProject)
+						 .map(cLanguageServerProvider::isEnabledFor)
+						 .orElse(Boolean.FALSE);
+			} else if (receiver instanceof DocumentSymbolWithFile) {
+				// called to enable the LS based CSymbolsContentProvider:
+				return true;
+			} else if (receiver instanceof IProject) {
+				return cLanguageServerProvider.isEnabledFor((IProject) receiver);
+			}
+			return false;
+		}
+		return false;
+	}
+	
+	private IFile getFileHandle(URI uri) {
+		if (uri == null) {
+			return null;
+		}
+		if (FILE_SCHEME.equals(uri.getScheme())) {
```

Comment:
Why do we need this check?

Reply:
Since the uri comes from LSP4E I've looked how they implemented the uri to file conversion. I'am doing it here the same way as they have done it. To be frank I've not questioning it yet. But you are right we can try:
`LspPlugin.getDefault().getWorkspace().getRoot().findFilesForLocationURI(uri);`
and if it fails we do: `Adapters.adapt(uri.toString(), IFile.class, true);`